pkg/ksym/ksym.go

```diff
@@ -140,66 +154,79 @@ func (c *cache) Resolve(addrs map[uint64]struct{}) (map[uint64]string, error) {
 	return res, nil
 }
 
-func unsafeString(b []byte) string {
-	return *((*string)(unsafe.Pointer(&b)))
-}
-
-// ksym reads /proc/kallsyms and resolved the addresses to their respective
-// function names. The addrs parameter must be sorted as /proc/kallsyms is
-// sorted.
-func (c *cache) ksym(addrs []uint64) ([]string, error) {
+// loadKsyms reads /proc/kallsyms and stores the start address for every function
+// names, sorted by the start address.
+func (c *cache) loadKsyms() error {
 	fd, err := c.fs.Open("/proc/kallsyms")
 	if err != nil {
-		return nil, err
+		return err
 	}
 	defer fd.Close()
 
-	res := make([]string, 0, len(addrs))
-
 	s := bufio.NewScanner(fd)
-	lastSym := ""
+	symbol := ""
 
 	for s.Scan() {
-		l := s.Bytes()
+		line := s.Bytes()
 
-		curAddr, err := strconv.ParseUint(unsafeString(l[:16]), 16, 64)
```

Comment:
I think here `unsafeString` is used to prevent additional allocations for the interim string! This might help https://chris124567.github.io/2021-06-21-go-performance/ (I couldn't find a better resource, I need to dig more!)

Reply:
Realised it was to avoid allocations, but thought that the unsafe wasn't worth it for the performance improvements. While the improvement is far from minimal, it's trumped by the rest of the code that does perform way more allocations and uses more CPU cycles to operate.

Added some basic benchmarks in the new commit:
```
[javierhonduco@fedora parca-agent]$ go test -v -bench=. -benchmem github.com/parca-dev/parca-agent/pkg/ksym
=== RUN   TestEnsureKsymSizeDoesNotGrow
--- PASS: TestEnsureKsymSizeDoesNotGrow (0.00s)
=== RUN   TestKsym
--- PASS: TestKsym (0.00s)
goos: linux
goarch: amd64
pkg: github.com/parca-dev/parca-agent/pkg/ksym
cpu: Intel(R) Core(TM) i7-8700K CPU @ 3.70GHz
BenchmarkStringParsing
BenchmarkStringParsing-12        2184770               727.6 ns/op           384 B/op         24 allocs/op
BenchmarkUnsafeString
BenchmarkUnsafeString-12        106100638               11.14 ns/op            0 B/op          0 allocs/op
PASS
ok      github.com/parca-dev/parca-agent/pkg/ksym       3.853s
```